sensorprocess/sensorprocess_test.go

```diff
@@ -236,18 +236,91 @@ func TestAddSensorReadingLiveReadings(t *testing.T) {
 			return errUnknown
 		}
 
-		timeToSleep := addSensorReadingFromLiveReadings(context.Background(), reading, readingTimestamp, config)
+		timeToSleep := addSensorReadingOnline(context.Background(), reading, readingTimestamp, config)
 		test.That(t, timeToSleep, test.ShouldBeGreaterThan, 0)
 		test.That(t, timeToSleep, test.ShouldBeLessThanOrEqualTo, config.LidarDataRateMsec)
 	})
 }
 
+func onlineModeTestHelper(
+	ctx context.Context,
+	t *testing.T,
+	config Config,
+	cf cartofacade.Mock,
+	cam string,
+) {
+	logger := golog.NewTestLogger(t)
+	liveSensor, err := s.NewLidar(context.Background(), s.SetupDeps(cam), cam, logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	var calls []addSensorReadingArgs
+	cf.AddSensorReadingFunc = func(
+		ctx context.Context,
+		timeout time.Duration,
+		sensorName string,
+		currentReading []byte,
+		readingTimestamp time.Time,
+	) error {
+		args := addSensorReadingArgs{
+			timeout:          timeout,
+			sensorName:       sensorName,
+			currentReading:   currentReading,
+			readingTimestamp: readingTimestamp,
+		}
+		calls = append(calls, args)
+		if len(calls) == 1 {
+			return errUnknown
+		}
+		if len(calls) == 2 {
+			return cartofacade.ErrUnableToAcquireLock
+		}
+		return nil
+	}
+
+	config.CartoFacade = &cf
+	config.Lidar = liveSensor
+	config.LidarName = liveSensor.Name
+
+	jobDone := addSensorReading(ctx, config, true)
+	test.That(t, len(calls), test.ShouldEqual, 1)
+	test.That(t, jobDone, test.ShouldBeFalse)
+
+	jobDone = addSensorReading(ctx, config, true)
+	test.That(t, len(calls), test.ShouldEqual, 2)
+	test.That(t, jobDone, test.ShouldBeFalse)
+
+	jobDone = addSensorReading(ctx, config, true)
+	test.That(t, len(calls), test.ShouldEqual, 3)
+	test.That(t, jobDone, test.ShouldBeFalse)
+
+	for i, call := range calls {
+		t.Logf("call %d", i)
+		test.That(t, call.sensorName, test.ShouldResemble, cam)
+		// the lidar test fixture happens to always return the same pcd currently
+		// in reality it could be a new pcd every time
+		test.That(t, call.currentReading, test.ShouldResemble, expectedPCD)
+		test.That(t, call.timeout, test.ShouldEqual, config.Timeout)
+	}
+
+	if cam == "good_lidar" {
+		test.That(t, calls[0].readingTimestamp.Before(calls[1].readingTimestamp), test.ShouldBeTrue)
+		test.That(t, calls[1].readingTimestamp.Before(calls[2].readingTimestamp), test.ShouldBeTrue)
+	} else if cam == "replay_lidar" {
+		readingTime, err := time.Parse(time.RFC3339Nano, s.TestTime)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, calls[0].readingTimestamp.Equal(readingTime), test.ShouldBeTrue)
+	} else {
+		t.Errorf("no timestamp tests provided for %v", cam)
+	}
```

Comment:
This is the only part that was changed so that the test could be generalized across replay & live lidars